A88-xds-data-error-handling.md

```diff
@@ -0,0 +1,451 @@
+  to also report that the client has lost contact with the xDS server.
+
+### Handling Data Errors
+
+As described above, the behavior that we want on data errors depends
+on whether we can depend on the control plane to properly alert human
+operators when these errors occur.  To configure this, we will introduce a
+new server feature in the bootstrap config called "fail_on_data_errors",
+which indicates that we cannot depend on the control plane to properly
+alert human operators.
+
+This server feature affects all data errors, which are the following cases:
+- NACKs
+- resource deletions in LDS and CDS
+- [xRFC TP3] errors with status NOT_FOUND or PERMISSION_DENIED
+
+If one of these errors occurs and there is no previously cached version
+of the resource, then the XdsClient will call the watchers'
+`OnResourceChanged()` method with an appropriate error.
+
+If one of these errors occurs when there IS a previously cached version
+of the resource, then the behavior depends on whether the
+"fail_on_data_errors" server feature is present in the bootstrap config.
+If the server feature is present, then the XdsClient will remove the
+previous version of the resource from its cache and will call the
+watchers' `OnResourceChanged()` method with an appropriate error.  If
+the server feature is NOT present, then the XdsClient will retain the
+previous version of the resource in its cache and will call the
+watchers' `OnAmbientError()` method with an appropriate error.
+
+### Handling Transient Errors
+
+Transient errors include the following cases:
+- the xDS channel reporting TRANSIENT_FAILURE or the ADS stream failing
+  without having received a response from the server, as described in
+  [gRFC A57]
+- [xRFC TP3] errors with any status *except* NOT_FOUND or PERMISSION_DENIED
+
+When one of these errors occurs, if we do NOT have a previously cached
+version of the resource, the XdsClient will call the watchers'
+`OnResourceChanged()` method to notify them of the error.  If we DO have
+a previously cached version of the resource, the XdsClient will instead
+call the watchers' `OnAmbientError()` method to notify them of the
+error.
+
+### Support for Errors Provided by the xDS server as per [xRFC TP3]
+
+We will add support for the new response fields added in [xRFC TP3].
+There will be no configuration telling gRPC to use these fields; we will
+unconditionally use them if they are present.
+
+When we receive an error for a given resource, we will do the following:
+1. Cancel the resource timer, if any.
+2. Set the cache entry's status to `RECEIVED_ERROR` and record the error,
+   so that that information can be reported via CSDS.
+3. If the status code is NOT_FOUND or PERMISSION_DENIED and the
+   "fail_on_data_errors" server feature is present in the bootstrap
+   config, delete the existing cached resource, if any.
+4. If there is an existing cached resource, call the watchers'
```

Comment:
For (3) you mention "fail_on_data_errors", but that isn't mentioned in (4), even though it would also affect whether or not to use `OnAmbientError`

Reply:
We don't actually need to check it here.  I've updated the wording to make this clear.